### PR TITLE
add test_spi.py

### PR DESCRIPTION
**Pull request by mntng**

### CODE REVIEW

**Comment by jordens** [comment]
Please add proper commit messages stating what was done and if possible why it was done and how it was done. The ones auto-generated by github are insufficient.

File: artiq/test/coredevice/test_spi.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                      0*spi.SPI_LSB_FIRST | 0*spi.SPI_HALF_DUPLEX)
+                      
+class CARDTest(EnvExperiment):
+    def build(self):
+        self.setattr_device("core")
+        self.setattr_device("spi_mmc")
+
+    @kernel
+    def run(self):  
+
```
Comment (sbourdeauducq) [inline artiq/test/coredevice/test_spi.py]: Remove that blank line.

File: artiq/test/coredevice/test_spi.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        self.set_dataset("sd_response", sd_response)
+    
+            
+ 
+class SDTest(ExperimentCase):
+    def test(self):
+        self.execute(CARDTest)
+        self.assertTrue(self.dataset_mgr.get("sd_response"))
+                
+
```
Comment (sbourdeauducq) [inline artiq/test/coredevice/test_spi.py]: One too many blank lines.

File: artiq/test/coredevice/test_spi.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+import os, unittest
```
Comment (sbourdeauducq) [inline artiq/test/coredevice/test_spi.py]: ``os`` is not used, don't import it.

File: artiq/test/coredevice/test_spi.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import os, unittest
+from artiq.experiment import *
+from artiq.test.hardware_testbench import ExperimentCase
+from artiq.language.core import (kernel, portable, delay_mu, delay)
+from artiq.language.units import ns, us
```
Comment (sbourdeauducq) [inline artiq/test/coredevice/test_spi.py]: ``ns`` is also not used.

File: artiq/test/coredevice/test_spi.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
+import os, unittest
+from artiq.experiment import *
+from artiq.test.hardware_testbench import ExperimentCase
+from artiq.language.core import (kernel, portable, delay_mu, delay)
```
Comment (sbourdeauducq) [inline artiq/test/coredevice/test_spi.py]: ``portable`` is not used.

File: artiq/test/coredevice/test_spi.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+from artiq.language.units import ns, us
+from artiq.coredevice import spi
+
+
+_SDCARD_SPI_CONFIG = (0*spi.SPI_OFFLINE | 0*spi.SPI_CS_POLARITY |
+                      0*spi.SPI_CLK_POLARITY | 0*spi.SPI_CLK_PHASE |
+                      0*spi.SPI_LSB_FIRST | 0*spi.SPI_HALF_DUPLEX)
+                      
+class CARDTest(EnvExperiment):
```
Comment (sbourdeauducq) [inline artiq/test/coredevice/test_spi.py]: ``CARD`` → ``Card``

File: artiq/examples/master/device_db.py [hunk 7]
```diff
@@ -137,6 +137,13 @@
         "class": "SPIMaster",
         "arguments": {"channel": 23}
     },
+    
+    "spi_mmc": { 
+        "type": "local",
+        "module": "artiq.coredevice.spi",
+        "class": "SPIMaster",
+        "arguments": {"channel": 26}
+    },
```
Comment (sbourdeauducq) [inline artiq/examples/master/device_db.py]: This change (and others) is already merged. Please don't submit them again (try ``git pull``).

File: artiq/test/coredevice/test_spi.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                sd_response = True
+                break
+        self.set_dataset("sd_response", sd_response)
+            
+ 
+class SDTest(ExperimentCase):
+    def test(self):
+        self.execute(CARDTest)
+        self.assertTrue(self.dataset_mgr.get("sd_response"))
+
```
Comment (sbourdeauducq) [inline artiq/test/coredevice/test_spi.py]: Now you have some spurious spaces at the end of your file... (and e.g. one line 56)

Comment (mntng) [inline artiq/test/coredevice/test_spi.py]: ............???

Comment (sbourdeauducq) [inline artiq/test/coredevice/test_spi.py]: Your file contains space characters (" ") that should not be there.

**Comment by sbourdeauducq** [comment]
The test does not pass.
http://buildbot.m-labs.hk/builders/artiq/builds/1636/steps/python_unittest_2/logs/stdio
```

======================================================================
FAIL: test (coredevice.test_spi.SDTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/lib/buildbot/slaves/debian-stretch-amd64-1/artiq/build/artiq/test/coredevice/test_spi.py", line 61, in test
    self.assertTrue(self.dataset_mgr.get("sd_response"))
AssertionError: False is not true

----------------------------------------------------------------------
```

**Comment by jbqubit** [comment]
@mntng Please include narrative describing what test_spi.py is meant to test. That is, what was missing from ARTIQ prior to your commit and how does your commit address it.

**Pull request merged**